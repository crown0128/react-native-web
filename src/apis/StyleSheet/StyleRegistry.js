/**
 * WARNING: changes to this file in particular can cause significant changes to
 * the results of render performance benchmarks.
 */
import createReactDOMStyle from './createReactDOMStyle';
import flattenArray from '../../modules/flattenArray';
import flattenStyle from './flattenStyle';
import I18nManager from '../I18nManager';
import mapKeyValue from '../../modules/mapKeyValue';
import { prefixInlineStyles } from '../../modules/prefixStyles';
import ReactNativePropRegistry from '../../modules/ReactNativePropRegistry';
import StyleManager from './StyleManager';

const createCacheKey = id => {
  const prefix = I18nManager.isRTL ? 'rtl' : 'ltr';
  return `${prefix}-${id}`;
};

const classListToString = list => list.join(' ').trim();

class StyleRegistry {
  constructor() {
    this.cache = {};
    this.styleManager = new StyleManager();
  }

  getStyleSheetHtml() {
    return this.styleManager.getStyleSheetHtml();
  }

  /**
   * Registers and precaches a React Native style object to HTML class names
   */
  register(flatStyle) {
    const id = ReactNativePropRegistry.register(flatStyle);
    const key = createCacheKey(id);
    const style = createReactDOMStyle(flatStyle);
    const classList = mapKeyValue(style, (prop, value) => {
      if (value != null) {
        return this.styleManager.setDeclaration(prop, value);
      }
    });
    const className = classList.join(' ').trim();
    this.cache[key] = { classList, className };
    return id;
  }

  /**
   * Resolves a React Native style object to DOM attributes
   */
  resolve(reactNativeStyle) {
    if (!reactNativeStyle) {
      return undefined;
    }

    // fast and cachable
    if (typeof reactNativeStyle === 'number') {
      const key = createCacheKey(reactNativeStyle);
      return this._resolveStyleIfNeeded(key, reactNativeStyle);
    }

    // resolve a plain RN style object
    if (!Array.isArray(reactNativeStyle)) {
      return this._resolveStyle(reactNativeStyle);
    }

    // flatten the style array
    // cache resolved props when all styles are registered
    // otherwise fallback to resolving
    const flatArray = flattenArray(reactNativeStyle);
    let isArrayOfNumbers = true;
    for (let i = 0; i < flatArray.length; i++) {
      if (typeof flatArray[i] !== 'number') {
        isArrayOfNumbers = false;
        break;
      }
    }
    const key = isArrayOfNumbers ? createCacheKey(flatArray.join('-')) : null;
    return this._resolveStyleIfNeeded(key, flatArray);
  }

  /**
   * Resolves a React Native style object to DOM attributes, accounting for
   * the existing styles applied to the DOM node.
   *
   * To determine the next style, some of the existing DOM state must be
   * converted back into React Native styles.
   */
  resolveStateful(rnStyleNext, { classList: rdomClassList, style: rdomStyle }) {
    // Convert the DOM classList back into a React Native form
    // Preserves unrecognized class names.
    const { classList: rnClassList, style: rnStyle } = rdomClassList.reduce(
      (styleProps, className) => {
        const { prop, value } = this.styleManager.getDeclaration(className);
        if (prop) {
          styleProps.style[prop] = value;
        } else {
          styleProps.classList.push(className);
        }
        return styleProps;
      },
      { classList: [], style: {} }
    );

    // Create next DOM style props from current and next RN styles
    const { classList: rdomClassListNext, style: rdomStyleNext } = this.resolve([
      rnStyle,
      rnStyleNext
    ]);

    // Next class names take priority over current inline styles
    const style = { ...rdomStyle };
    rdomClassListNext.forEach(className => {
      const { prop } = this.styleManager.getDeclaration(className);
      if (style[prop]) {
        style[prop] = '';
      }
    });

    // Next inline styles take priority over current inline styles
    Object.assign(style, rdomStyleNext);

    // Add the current class names not managed by React Native
    const className = classListToString(rdomClassListNext.concat(rnClassList));

    return { className, style };
  }

  /**
   * Resolves a React Native style object
   */
  _resolveStyle(reactNativeStyle) {
    const domStyle = createReactDOMStyle(flattenStyle(reactNativeStyle));

    const props = Object.keys(domStyle).reduce(
      (props, styleProp) => {
        const value = domStyle[styleProp];
        if (value != null) {
          const className = this.styleManager.getClassName(styleProp, value);
          if (className) {
            props.classList.push(className);
          } else {
            if (!props.style) {
              props.style = {};
            }
            // 4x slower render
            props.style[styleProp] = value;
          }
        }
        return props;
      },
      { classList: [] }
    );

    props.className = classListToString(props.classList);
    if (props.style) {
      props.style = prefixInlineStyles(props.style);
    }
    return props;
  }

  /**
  * Caching layer over 'resolveStyle'
   */
  _resolveStyleIfNeeded(key, style) {
    if (key) {
      if (!this.cache[key]) {
        // slow: convert style object to props and cache
        this.cache[key] = this._resolveStyle(style);
      }
      return this.cache[key];
    }
    return this._resolveStyle(style);
  }
}

module.exports = StyleRegistry;
