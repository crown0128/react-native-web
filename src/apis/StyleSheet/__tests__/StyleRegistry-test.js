/* eslint-env jasmine, jest */

import StyleRegistry from '../StyleRegistry';

let styleRegistry;

describe('apis/StyleSheet/StyleRegistry', () => {
  beforeEach(() => {
    styleRegistry = new StyleRegistry();
  });

  test('register', () => {
    const style = { opacity: 0 };
    const id = styleRegistry.register(style);
    expect(typeof id === 'number').toBe(true);
  });

  describe('resolve', () => {
    const styleA = { borderWidth: 0, borderColor: 'red', width: 100 };
    const styleB = {
      position: 'absolute',
      left: 50,
      pointerEvents: 'box-only'
    };
    const styleC = { width: 200 };

    const testResolve = (a, b, c) => {
      // no common properties, different resolving order, same result
      const resolve1 = styleRegistry.resolve([a, b]);
      expect(resolve1).toMatchSnapshot();
      const resolve2 = styleRegistry.resolve([b, a]);
      expect(resolve1).toEqual(resolve2);

      // common properties, different resolving order, different result
      const resolve3 = styleRegistry.resolve([a, b, c]);
      expect(resolve3).toMatchSnapshot();
      const resolve4 = styleRegistry.resolve([c, a, b]);
      expect(resolve4).toMatchSnapshot();
      expect(resolve3).not.toEqual(resolve4);
    };

    test('with register, resolves to className', () => {
      const a = styleRegistry.register(styleA);
      const b = styleRegistry.register(styleB);
      const c = styleRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('with register, resolves to mixed', () => {
      const a = styleA;
      const b = styleRegistry.register(styleB);
      const c = styleRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('without register, resolves to inline styles', () => {
      testResolve(styleA, styleB, styleC);
    });
  });

  test('resolveStateful', () => {
    // generate a classList to act as pre-existing DOM state
    const mockStyle = styleRegistry.register({
      borderWidth: 0,
      borderColor: 'red',
      width: 100
    });
    const { classList: domClassList } = styleRegistry.resolve(mockStyle);
    domClassList.unshift('external-className');
    expect(domClassList).toMatchSnapshot();

    // test result
    const result = styleRegistry.resolveStateful({ borderWidth: 1, opacity: 1 }, domClassList);
    expect(result).toMatchSnapshot();
  });
});
