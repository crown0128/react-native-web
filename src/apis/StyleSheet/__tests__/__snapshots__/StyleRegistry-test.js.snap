exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 1`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-b8lwoo",
  "style": Object {},
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 2`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-l0gwng",
  "style": Object {},
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 3`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-b8lwoo",
  "style": Object {},
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 1`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 2`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-l0gwng",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 3`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 1`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 2`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "200px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 3`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful 1`] = `
Array [
  "external-className",
  "rn-borderTopColor-1gxhl28",
  "rn-borderRightColor-knoah9",
  "rn-borderBottomColor-1ani3fp",
  "rn-borderLeftColor-ribj9x",
  "rn-borderTopWidth-13yce4e",
  "rn-borderRightWidth-fnigne",
  "rn-borderBottomWidth-ndvcnb",
  "rn-borderLeftWidth-gxnn5r",
  "rn-width-b8lwoo",
]
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful 2`] = `
Object {
  "className": "rn-borderBottomColor-1ani3fp rn-borderBottomWidth-ndvcnb rn-borderLeftColor-ribj9x rn-borderLeftWidth-gxnn5r rn-borderRightColor-knoah9 rn-borderRightWidth-fnigne rn-borderTopColor-1gxhl28 rn-borderTopWidth-13yce4e rn-width-b8lwoo external-className",
  "style": Object {
    "borderBottomColor": null,
    "borderBottomWidth": null,
    "borderLeftColor": null,
    "borderLeftWidth": null,
    "borderRightColor": null,
    "borderRightWidth": null,
    "borderTopColor": null,
    "borderTopWidth": null,
    "opacity": 1,
    "width": null,
  },
}
`;
